### PR TITLE
Updates LLT to inplace decomposition per eigen 3.3 doc

### PR DESCRIPTION
**Pull request by rtrangucci**
#### Submission Checklist

- [ ] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

Inplace LLT saves memory and one copy. Addresses #359 

#### Intended Effect:

Reduction in memory, reduction in copies.

#### How to Verify:

 ./runTests.py test/unit/math/rev/mat/fun/cholesky_decompose_test.cp

#### Side Effects:

None

#### Documentation:

None

#### Reviewer Suggestions: 

@bob-carpenter 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Rob Trangucci

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: stan/math/rev/mat/fun/cholesky_decompose.hpp [hunk 1]
```diff
@@ -233,8 +233,6 @@ namespace stan {
      * Internally calls llt rather than using cholesky_decompose in order to
      * use selfadjointView<Lower> optimization.
```
Comment (bob-carpenter) [inline stan/math/rev/mat/fun/cholesky_decompose.hpp]: Is this doc broken now?  You've removed the `selfadjointView<Lower>` bit of the code.

File: stan/math/rev/mat/fun/cholesky_decompose.hpp [hunk 2]
```diff
@@ -246,10 +244,8 @@ namespace stan {
       check_symmetric("cholesky_decompose", "A", A);
 
       Eigen::Matrix<double, -1, -1> L_A(value_of_rec(A));
-      Eigen::LLT<Eigen::MatrixXd> L_factor
-        = L_A.selfadjointView<Eigen::Lower>().llt();
+      Eigen::LLT<Eigen::Ref<Eigen::MatrixXd> > L_factor(L_A);
```
Comment (bob-carpenter) [inline stan/math/rev/mat/fun/cholesky_decompose.hpp]: Is there a reason we're no longer using a selfadjointView of `L_A`?   Have the input requirements changed now that we use all of it?

Comment (rtrangucci) [inline stan/math/rev/mat/fun/cholesky_decompose.hpp]: Good question! I'm not sure if I can do a Ref to a triangular view of a matrix. I'll see if I can figure it out.

Comment (rtrangucci) [inline stan/math/rev/mat/fun/cholesky_decompose.hpp]: I used the selfadjointView<Eigen::Lower>() to cut down on the copy when L_factor was instantiated, but I'm fairly certain that LLT defaults to a lower triangular view. In any case, the inplace decomposition obviates the need for any selfadjointView() instantiation here. I've added an explicit Eigen::Lower template argument, but that's the default for the class.

File: stan/math/rev/mat/fun/cholesky_decompose.hpp (hunk 1 above)
Comment (rtrangucci) [inline stan/math/rev/mat/fun/cholesky_decompose.hpp]: Fixed the doc. Thanks for the review.

**Pull request merged**